fix: Fix concurrency issue in message call processors

kselveliev opened the pull request:
**Description**:
Currently making simultaneous calls of decimals function with multiple threads against erc tokens with different decimals will result in mixed up results. If one token has 10 decimals the other has 8 decimals, when we expect the result to be 10 it might turn out to be 8 and vice versa. This is because our MessageCallProcessors - `MirrorEvmMessageCallProcessor` and `MirrorEvmMessageCallProcessorV30` are singleton beans that extend hedera-evm `HederaEvmMessageCallProcessor` which maintains state with 
```
protected long gasRequirement;
protected Bytes output;
```
Multiple threads calling `executeHederaPrecompile()` will modify simultaneously shared  `gasRequirement` and `output` fields and mix up the results.


This Pr modifies :
`MirrorEvmMessageCallProcessor` override `executeHederaPrecompile()` define `gasRequirement` and `output` as local variables within the executeHederaPrecompile to ensure thread safety.
 
`MirrorEvmMessageCallProcessorV30` override `executeHederaPrecompile()` define `gasRequirement` and `output` as local variables within the executeHederaPrecompile to ensure thread safety.


**Related issue(s)**:

Fixes #7572 

**Notes for reviewer**:
<!-- Provide logs, performance numbers or screenshots of the new functionality -->

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

kselveliev commented:
Not entirely sure if this issue should be handled here or could even be changed in hedera-evm in services.
What do you think @steven-sheehy

steven-sheehy commented:
> Not entirely sure if this issue should be handled here or could even be changed in hedera-evm in services. What do you think @steven-sheehy

Yes, I think ultimately we want this fixed upstream. Talked to Michael T and he said making it stateless sounds reasonable and seems like it's not used in modularized. It looks like we should split `executeHederaPrecompile()` into two methods with the first part returning the gas and output as `Pair<Long, Bytes>` and the second part doing the tracing and frame update. Maybe for now we can copy the `HederaEvmMessageCallProcessor` and make that change locally until we get the updated evm library with the upstream fix.

IvanKavaldzhiev commented:
I'm not sure it's worth to modify `HederaEvmMessageCallProcessor` inside `hedera-evm`, which would be soon deprecated. At a given point we would replace this library with modules from the modularized repo. Also, this behaviour shouldn't be an issue in services, since transactions there are executed synchronously one by one.

IvanKavaldzhiev reviewed (dismissed):
LGTM

kselveliev commented:
> > Not entirely sure if this issue should be handled here or could even be changed in hedera-evm in services. What do you think @steven-sheehy
> 
> Yes, I think ultimately we want this fixed upstream. Talked to Michael T and he said making it stateless sounds reasonable and seems like it's not used in modularized. It looks like we should split `executeHederaPrecompile()` into two methods with the first part returning the gas and output as `Pair<Long, Bytes>` and the second part doing the tracing and frame update. Maybe for now we can copy the `HederaEvmMessageCallProcessor` and make that change locally until we get the updated evm library with the upstream fix.

I refactored the code, split the `executeHederaPrecompile()` into 2 methods one calculating gas and output and the other doing the tracing and frame update. Extracted common methods in a base abstract class to avoid code duplication

0xivanov reviewed (dismissed):
LGTM

steven-sheehy commented on hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/contracts/execution/MirrorEvmMessageCallProcessorV30.java:
```diff
@@ -35,4 +39,15 @@ public MirrorEvmMessageCallProcessorV30(
     }
+
+    @Override
+    protected void executeHederaPrecompile(
```
Is this method identical in each sub class? If so can move to parent

kselveliev replied:
Identical indeed. I extracted it in parent class.

steven-sheehy reviewed (approved):
LGTM
